appengine/flexible/TwelveFactor/Services/CloudStorageFileProvider.cs

```diff
@@ -0,0 +1,259 @@
+        /// bucket name and object name.  </summary>
+        ObjectPath SplitObjectPath(string path)
+        {
+            // Accept paths in a variety of forms:
+            //   bucket/a/b/c
+            //   /bucket/a/b/c
+            //   gs://bucket/a/b/c
+            string gsPrefix = "gs:/";
+            if (path.ToLower().Substring(0, gsPrefix.Length) == gsPrefix) {
+                path = path.Substring(gsPrefix.Length);
+            }
+            if ('/' == path.FirstOrDefault()) {
+                path = path.Substring(1);
+            }
+            string[] fragments = path.Split(new [] {'/'}, 2);
+            return new ObjectPath()
+            {
+                BucketName = fragments.FirstOrDefault(),
+                ObjectName = fragments.LastOrDefault()
+            };
+        }
+
+        /// <summary> Gets the Object at the path.  If it's not found,
+        /// doesn't throw an exception; instead returns null. </summary>
+        Google.Apis.Storage.v1.Data.Object GetObject(ObjectPath path) {
+            try 
+            {
+                return _storage.GetObject(path.BucketName, 
+                    path.ObjectName);
+            }
+            catch (Google.GoogleApiException e)
+            when (e.HttpStatusCode == HttpStatusCode.NotFound) 
+            {
+                _logger.LogWarning("{0}/{1} not found.", path.BucketName,
+                    path.ObjectName);
+                return null;
+            }
+        }
+
+        /// <summary> Watch a file for changes.
+        public IChangeToken Watch(string filter)
+        {
+            if (_pollingInterval == null) {
+                return null;
+            }
+            _logger.LogDebug("Watching {0}", filter);
+            var path = SplitObjectPath(filter);
+            var token = new CloudStorageChangeToken();
+            var obj = GetObject(path);
+            long? generation = obj?.Generation;
+            Task.Run(async () => {
+                while (true) {
+                    await Task.Delay(_pollingInterval.Value);
+                    try
+                    {
+                        obj = GetObject(path);
+                        if (generation != obj?.Generation) {
+                            token.HasChanged = true;
+                            return;
+                        }
+                    } 
+                    catch (Exception e) 
+                    {
+                        _logger.LogError(0, e, "Exception while watching {0}",
+                            filter);
+                    }
+                }
+            });
+            return token;            
+        }            
+    }
+
+    class CloudStorageChangeToken : IChangeToken
+    {
+        List<ChangeCallback> _callbacks = new List<ChangeCallback>();
+        bool _hasChanged = false;
+
+        public bool HasChanged { 
+            get { return _hasChanged; }
+            set {
+                _hasChanged = true;

```

Comment:
You should verify that `value` is indeed `true`. Maybe this should be a readonly property with a method for teh change? The `IChangeToken` interface only specifies a getter.